### ISSUE TITLE
[new-lint] if let Some... else with locked mutex

### ISSUE THREAD

**Issue by DevinR528** [rust-lang/rust-clippy#5219]
I was working on a project and did something like this
```rust
let mutex: Mutex<Option<...>>;
if let Some(thing) = mutex.lock().unwrap() {
    do_thing();
} else {
    mutex.lock();
}
```
It just sat running and it took me a bit to realize the mutex is locked for the entire scope of the `if` and `else`. The suggestion could be something like
```
warning: locking a `Mutex` twice in the same scope will (fail/freeze/deadlock)
if let Some(thing) = mutex.lock().unwrap() {  <=== lock starts here
    do_thing();                               <=
} else {                                      <=
    mutex.lock();      <= second lock         <= 
}                                             <= lock is droped
```
or a suggestion to
```
warning: move lock out of `if let...else`
let locked = mutex.lock().unwrap();
if let Some(thing) = locked {
    do_thing();
} else {
    do_something_else(locked);
}  
```
I would be happy to start a PR if this sounds good to anyone else!

**Comment by DevinR528** [comment]
I should be using the latest nightly for testing a PR right? Is there a branch that is rustc 1.43.0-nightly current to start working from or is this a case where I should start with an update PR? @matthiaskrgr where did you start your PR from your seems to be current with the removal of `rustc_infer`?

**Comment by flip1995** [comment]
Clippy is compiled with the latest master commit of rustc, not with the latest nightly. (I think the 2020-02-25 nightly should work right now, though)

The `setup-toolchain.sh` script will set up the master toolchain for you.

**Comment by viruscamp** [comment]
failed on
```rust
    let vec_mutex = Mutex::new(vec![2,3]);
    if let Some(num) = vec_mutex.lock().unwrap().pop() {
        if num == 2 {
            vec_mutex.lock().unwrap().push(4);
        }
        println!("got {num}");
    };
```
